### PR TITLE
Remotes/origin/workloadconsistent

### PR DESCRIPTION
**Pull request by soumyapattnaik**

### CODE REVIEW

File: VMBackup/main/handle.py [hunk 1]
```diff
@@ -338,13 +347,13 @@ def daemon():
...
             backup_logger.log('timespan is ' + str(timespan) + ' ' + str(total_span_in_seconds))
 
-            if total_span_in_seconds > MAX_TIMESPAN :
-                error_msg = "CRP timeout limit has reached, will not take snapshot."
-                errMsg = error_msg
-                hutil.SetExtErrorCode(ExtensionErrorCodeHelper.ExtensionErrorCodeEnum.FailedGuestAgentInvokedCommandTooLate)
-                temp_result=CommonVariables.FailedGuestAgentInvokedCommandTooLate
-                temp_status= 'error'
-                exit_with_commit_log(temp_status, temp_result,error_msg, para_parser)
+#            if total_span_in_seconds > MAX_TIMESPAN :
```
Comment (soumyapattnaik) [inline VMBackup/main/handle.py:350]: this is temporary. i will revert it in next commit

File: VMBackup/main/workloadPatch/WorkloadPatch.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+class WorkloadPatch:
+    def __init__(self, workload_name, logger):
+        self.logger = logger
+        self.name = workload_name
+        self.command = "/usr/bin/"
+        self.dbnames = []
+        self.login_path = ""
+        self.ipc_folder = None
+        self.error_details = []
+        self.enforce_slave_only = True
```
Comment (soumyapattnaik) [inline VMBackup/main/workloadPatch/WorkloadPatch.py:39]: make it 1

File: VMBackup/main/workloadPatch/WorkloadUtils/OracleUtility.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        oracleContainerName = subprocess.check_output(containerNameArgs, shell=True)
+        self.logger.log("Shrid: containerName- " + str(oracleContainerName))
+        print("Shrid: containerName- ", str(oracleContainerName))
+        
+        if "CDB$ROOT" in str(oracleContainerName):
+            self.logger.log("Shrid: containerName- In CDB$ROOT")
+            print("Shrid: containerName- In CDB$ROOT")
+            return True
+        else:
+            self.logger.log("Shrid: Pre- Error. Not in CDB$ROOT")
```
Comment (soumyapattnaik) [inline VMBackup/main/workloadPatch/WorkloadUtils/OracleUtility.py:47]: please remove shrid everywhere

Comment (soumyapattnaik) [reply VMBackup/main/workloadPatch/WorkloadUtils/OracleUtility.py:47]: @shridpant

File: VMBackup/main/workloadPatch/WorkloadUtils/logRestore.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+from time import sleep
+import subprocess
+from datetime import datetime
+import re
+try:
+    import ConfigParser as ConfigParsers
+except ImportError:
+    import configparser as ConfigParsers
+
+class Incremental:
```
Comment (soumyapattnaik) [inline VMBackup/main/workloadPatch/WorkloadUtils/logRestore.py:13]: @shridpant -rename to LogRestore also the classname to LogRestore

File: VMBackup/main/workloadPatch/WorkloadUtils/OracleUtility.py (hunk 3 above)
Comment (soumyapattnaik) [reply VMBackup/main/workloadPatch/WorkloadUtils/OracleUtility.py:47]: remove print statements

File: VMBackup/main/workloadPatch/WorkloadUtils/logRestore.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        regX = re.compile(r"\*\..+=.+")
+        parameterFile = open(self.parameterFilePath, 'r')
+        contents = parameterFile.read()
+        for match in regX.finditer(contents):
+            keyParameter = match.group().split('=')[0].lstrip('*\.')
+            valueParameter = [name.strip('\'') for name in match.group().split('=')[1].split(',')]
+            self.oracleParameter[keyParameter] = valueParameter
+        #print(self.oracleParameter)
+
+    #----Start Incremental Restore----#
```
Comment (soumyapattnaik) [inline VMBackup/main/workloadPatch/WorkloadUtils/logRestore.py:36]: function descriptive description

File: VMBackup/main/workloadPatch/WorkloadUtils/logRestore.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    def switchArchiveLogFiles(self, backupPath):
+        recoveryFileDest = self.oracleParameter['db_recovery_file_dest']
+        dbName = self.oracleParameter['db_name']
+        for location in recoveryFileDest:
+            os.system('rm -R -f '+ location + '/' + dbName[0] +'/archivelog')
+            os.system('cp -R -f ' + backupPath + '/archivelog ' + location + '/' + dbName[0] + '/archivelog')
+            os.system('chmod -R a+wrx '+ location +'/' + dbName[0] + '/archivelog')
+        print("Switched Archive Log Files")
+
+    def restore(self):
```
Comment (soumyapattnaik) [inline VMBackup/main/workloadPatch/WorkloadUtils/logRestore.py:54]: @shridpant - make this a meaningful name

File: VMBackup/main/workloadPatch/WorkloadUtils/logRestore.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+                #    print("    logbackup: config logbackup base location: ", self.baseLocation)
+                else:
+                    return None
+                if config.has_option("logbackup", 'crontabLocation'):
+                    self.crontabLocation = config.get("logbackup", 'crontabLocation')
+                #    print("    logbackup: config logbackup crontab location: ", self.crontabLocation)
+        else:
+            print("No matching workload config found")
+
+oracleIncremental = Incremental()
```
Comment (soumyapattnaik) [inline VMBackup/main/workloadPatch/WorkloadUtils/logRestore.py:96]: main code to handle this

File: VMBackup/main/workloadPatch/WorkloadUtils/logRestore.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
+import sys
```
Comment (soumyapattnaik) [inline VMBackup/main/workloadPatch/WorkloadUtils/logRestore.py:1]: add shebang

File: VMBackup/main/workloadPatch/logbackupPatch.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+import os
+from time import sleep
+try:
+    import ConfigParser as ConfigParsers
+except ImportError:
+    import configparser as ConfigParsers
+import subprocess
+
+class logbackup:
+    def __init__(self):
```
Comment (soumyapattnaik) [inline VMBackup/main/workloadPatch/logbackupPatch.py:12]: LogBackupPatch

File: VMBackup/main/logbackup.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
+import os
+import re
+import sys
+import subprocess
+import threading
+from workloadPatch.logbackupPatch import logbackup
+from time import sleep
+from datetime import datetime
+
+def parameterFileParser():
```
Comment (soumyapattnaik) [inline VMBackup/main/logbackup.py:10]: @shridpant - add main method and add shebang

Comment (soumyapattnaik) [reply VMBackup/main/logbackup.py:10]: move into Oracle utilities with OraLogBackup as name

File: VMBackup/main/logbackup.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
+import os
+import re
+import sys
+import subprocess
+import threading
+from workloadPatch.logbackupPatch import logbackup
+from time import sleep
+from datetime import datetime
+
```
Comment (soumyapattnaik) [inline VMBackup/main/logbackup.py:9]: @shridpant - insert a sample parameter file as comments for regex parsing

File: VMBackup/main/workloadPatch/WorkloadPatch.py [hunk 12]
```diff
@@ -0,0 +1,417 @@
...
+                self.error_details.append(ErrorDetail(CommonVariables.FailedWorkloadIPCDirectoryMissing, "IPC directory missing"))
+                return None
+            prescript = os.path.join(os.getcwd(), "main/workloadPatch/scripts/preMysqlMaster.sql")
+            arg = "sudo "+self.command+self.name+" "+self.cred_string+" -e\"set @timeout="+self.timeout+";set @outfile=\\\"\\\\\\\""+self.outfile+"\\\\\\\"\\\";source "+prescript+";\""
+            binary_thread = threading.Thread(target=self.thread_for_sql, args=[arg])
+            binary_thread.start()
+            waitForPreScriptCompletion()
+        elif 'oracle' in self.name.lower():
+            global preOracleStatus
+            preOracleStatus = self.databaseStatus()
```
Comment (soumyapattnaik) [inline VMBackup/main/workloadPatch/WorkloadPatch.py:119]: pre DB status and move it outside

File: VMBackup/main/workloadPatch/WorkloadPatch.py [hunk 13]
```diff
@@ -0,0 +1,417 @@
...
+            else:
+                self.logger.log("WorkloadPatch: Pre- Database not open.")
+                self.error_details.append(ErrorDetail(CommonVariables.FailedWorkloadDatabaseNotOpen, "Database not open"))
+                return None
+
+            self.logger.log("WorkloadPatch: Pre- Inside oracle pre")
+            preOracle = self.command + " -s / as sysdba @" + os.path.join(os.getcwd(), "main/workloadPatch/scripts/preOracleMaster.sql ")
+            args = ["su", "-", self.cred_string, "-c", preOracle]
+            process = subprocess.Popen(args)
+            while process.poll() == None:
```
Comment (soumyapattnaik) [inline VMBackup/main/workloadPatch/WorkloadPatch.py:131]: @shridpant - add timeout

File: VMBackup/main/workloadPatch/WorkloadPatch.py [hunk 14]
```diff
@@ -0,0 +1,417 @@
...
+
+    def waitForPreScriptCompletion(self):
+        if self.ipc_folder != None:
+            wait_counter = 5 
+            while len(self.child) == 0 and wait_counter > 0:
+                self.logger.log("child not created yet", True)
+                wait_counter--
+                sleep(2)
+            if wait_counter > 0:
+                self.logger.log("sql subprocess Created",True)
```
Comment (soumyapattnaik) [inline VMBackup/main/workloadPatch/WorkloadPatch.py:360]: check the status of child

File: VMBackup/main/workloadPatch/WorkloadPatch.py [hunk 15]
```diff
@@ -0,0 +1,417 @@
...
+                return None
+        
+    def timeoutDaemon(self):
+        global preDaemonThread
+        if 'oracle' in self.name.lower():
+            self.logger.log("WorkloadPatch: Inside oracle condition in timeout daemon")
+            preDaemonOracle = self.command + " -s / as sysdba @" + os.path.join(os.getcwd(), "main/workloadPatch/scripts/preOracleDaemon.sql ") + self.timeout
+            argsDaemon = ["su", "-", self.cred_string, "-c", preDaemonOracle]
+            preDaemonThread = threading.Thread(target=self.threadForTimeoutDaemon, args=[argsDaemon])
+            preDaemonThread.start()
```
Comment (soumyapattnaik) [inline VMBackup/main/workloadPatch/WorkloadPatch.py:385]: @shridpant - think of some failure cases here

File: VMBackup/main/workloadPatch/WorkloadPatch.py [hunk 16]
```diff
@@ -0,0 +1,417 @@
...
+        self.logger.log("WorkloadPatch: timeoutDaemon started for: " + self.timeout + " seconds")
+
+    def threadForTimeoutDaemon(self, args): 
+            global daemonProcess
+            daemonProcess = subprocess.Popen(args)
+            self.logger.log("WorkloadPatch: daemonProcess started")
+            while daemonProcess.poll() == None:
```
Comment (soumyapattnaik) [inline VMBackup/main/workloadPatch/WorkloadPatch.py:392]: @shridpant put a timeout

File: VMBackup/main/workloadPatch/WorkloadUtils/workload.conf [hunk 17]
```diff
@@ -0,0 +1,10 @@
+[workload]
+workload_name = oracle
```
Comment (soumyapattnaik) [inline VMBackup/main/workloadPatch/WorkloadUtils/workload.conf:2]: during final checkin all parameters needs to populated with comments

File: VMBackup/main/workloadPatch/WorkloadPatch.py (hunk 16 above)
Comment (shridpant) [reply VMBackup/main/workloadPatch/WorkloadPatch.py:392]: Done

File: VMBackup/main/workloadPatch/WorkloadPatch.py (hunk 13 above)
Comment (shridpant) [reply VMBackup/main/workloadPatch/WorkloadPatch.py:131]: Done

File: VMBackup/main/workloadPatch/WorkloadPatch.py (hunk 12 above)
Comment (shridpant) [reply VMBackup/main/workloadPatch/WorkloadPatch.py:119]: Done

File: VMBackup/main/logbackup.py (hunk 11 above)
Comment (shridpant) [reply VMBackup/main/logbackup.py:9]: Done

File: VMBackup/main/logbackup.py (hunk 10 above)
Comment (shridpant) [reply VMBackup/main/logbackup.py:10]: Done. It is named "OracleLogBackup.py"

Comment (shridpant) [reply VMBackup/main/logbackup.py:10]: Done. It is named "OracleLogBackup.py"

Comment (shridpant) [reply VMBackup/main/logbackup.py:10]: Done. It is named "OracleLogBackup.py"

File: VMBackup/main/workloadPatch/logbackupPatch.py (hunk 9 above)
Comment (shridpant) [reply VMBackup/main/workloadPatch/logbackupPatch.py:12]: Done

File: VMBackup/main/workloadPatch/WorkloadUtils/logRestore.py (hunk 8 above)
Comment (shridpant) [reply VMBackup/main/workloadPatch/WorkloadUtils/logRestore.py:1]: Done

File: VMBackup/main/workloadPatch/WorkloadUtils/logRestore.py (hunk 7 above)
Comment (shridpant) [reply VMBackup/main/workloadPatch/WorkloadUtils/logRestore.py:96]: Done

File: VMBackup/main/workloadPatch/WorkloadUtils/logRestore.py (hunk 6 above)
Comment (shridpant) [reply VMBackup/main/workloadPatch/WorkloadUtils/logRestore.py:54]: Done

File: VMBackup/main/workloadPatch/WorkloadUtils/logRestore.py (hunk 5 above)
Comment (shridpant) [reply VMBackup/main/workloadPatch/WorkloadUtils/logRestore.py:36]: Done

File: VMBackup/main/workloadPatch/WorkloadUtils/logRestore.py (hunk 4 above)
Comment (shridpant) [reply VMBackup/main/workloadPatch/WorkloadUtils/logRestore.py:13]: Done

File: VMBackup/main/workloadPatch/WorkloadUtils/OracleUtility.py (hunk 3 above)
Comment (shridpant) [reply VMBackup/main/workloadPatch/WorkloadUtils/OracleUtility.py:47]: Done

File: VMBackup/main/handle.py [hunk 18]
```diff
@@ -387,97 +392,131 @@ def daemon():
...
+                        g_fsfreeze_on = True
+                    else:
+                        g_fsfreeze_on = False
+                    freeze_snapshot(thread_timeout)
+                    workload_patch.post()
+                    workload_error = workload_patch.populateErrors()
+                    if workload_error != None and g_fsfreeze_on == False:
+                        run_status = 'error'
+                        run_result = workload_error.errorCode
+                        hutil.SetExtErrorCode(workload_error.errorCode)
```
Comment (bpramod) [inline VMBackup/main/handle.py:414]: shouldnt it succed with FS consistent backup ?

File: VMBackup/main/handle.py [hunk 19]
```diff
@@ -486,7 +529,7 @@ def daemon():
             error_msg = 'command is not correct'
             backup_logger.log(error_msg, True, 'Error')
     except Exception as e:
-        hutil.update_settings_file()
+        #hutil.update_settings_file()
```
Comment (soumyapattnaik) [inline VMBackup/main/handle.py:532]: to remove this

File: VMBackup/main/handle.py (hunk 18 above)
Comment (soumyapattnaik) [reply VMBackup/main/handle.py:414]: image

Comment (soumyapattnaik) [reply VMBackup/main/handle.py:414]: tested the above flows

File: VMBackup/main/handle.py [hunk 20]
```diff
@@ -387,97 +392,131 @@ def daemon():
                 else:
                     backup_logger.log("the logs blob uri is not there, so do not upload log.")
                 backup_logger.log('commandToExecute is ' + commandToExecute, True)
+                
+                workload_patch = WorkloadPatch.WorkloadPatch(backup_logger)
+                
+                if workload_patch.name != None:
```
[the remark this answers is not in the document]
Comment (soumyapattnaik) [reply VMBackup/main/handle.py:397]: done. if config parser fails we will skip pre post

**Pull request merged**